### DISCUSSION TITLE
Brave web browser on Mac and Linux got problem?

### DISCUSSION THREAD

**Discussion by pvanthony** [romcal/romcal#306] [Q&A] [answered]
Please note that I am not a programmer.

Did some javascript that uses romcal for a Tamil (Indian) website. It is working on Firefox and chrome on Linux, Mac, Windows, Android and Apple mobile. It does not seem to be working correctly on Brave web browser on Mac and Linux. Also Safari on MacOS is not working correctly. On Apple mobile iOS, Brave is not working correctly too.

Here is the website. https://iraivarthai.org/
All the javascript is in https://iraivarthai.org/packages/julian_custom.js

It is the first line below the menu bar. Something like, "Friday of the 27th week of Ordinary Time" is missing.

What did I miss doing? Need advise on how to solve this issue so that Brave and Safari will work correctly.

P.V.Anthony

**Comment by tukusejssirs** [comment]
@pvanthony, it does not work for me in Firefox Nightly 95.0a1 too.

In the [web console](https://developer.mozilla.org/en-US/docs/Tools/Web_Console) (nowadays, all browser have some kind of web console in the dev tools), it throws the following error:

```js
Uncaught (in promise) TypeError: can't access property "liturgicalColorNames", dailyReading[0] is undefined
julian_custom.js:201:21
<anonymous> https://iraivarthai.org/packages/julian_custom.js:201
```

`julian_custom.js:201:21` means:
- in file `julian_custom.js`;
- on line `201`;
- column `21` (i.e. 21st character on that line).

On L121, there is `if (dailyReading[0].liturgicalColorNames[0] == 'white'){`. The error means (in this case), that there is no property with a name `liturgicalColorNames` under `dailyReading[0]`, which _seems_ to mean that the definition of `deailyReading` is probably empty. You can see the output of `console.log()` function in the web console.

I would debug this by adding `console.log('dailyReading :', dailyReading)` after the definition of `dailyReading` (i.e. on L195) in order to see what does that variable contains. Then I would check the contents of `result` variable before and after L192.

So I tested it out. `dailyReading` is an array with no items (`dailyReading.length` is equal to `0`). `result` before L192 is an array of length 365 and after L192 is an array of length 122.

The issue lays on L192.

```js
// Original
var result = result.filter((day) => new Date(day.date).getMonth() >= today.getMonth() - 1 && new Date(day.date).getMonth() <= limitedMonths)

// Fixed
var result = result.filter((day) => new Date(day.date).getMonth() >= today.getMonth() && new Date(day.date).getMonth() <= limitedMonths)
```

The issues is that is subtract `1` from `today.getMonth()`, however I think you want to compare `new Date(day.date).getMonth()` with `today.getMonth()` (without subtraction), i.e. if they are equal.

After fixing this issue, `dailyReading` contains the today’s celebrations.

I invite you to check it out and come back to tell us if this also fixed the issue with Brave and Safari browsers.

**Reply by pvanthony** [reply]
@tukusejssirs thank you very much for the advice. It was very useful. Especially the console.log part.

The reason for the "-1" in the L192 is so that it will show the previous month as well. It is not perfect. Because in January it will not show December but that is ok. Trying my best to not load up the browser so that it will load as fast as possible.

Because of your help, I think the problem was the following code.
```
// If your locale is supported by Romcal, it will use it.
// If not, it will default to 'en'.
var currentLocale = navigator.language || navigator.userLanguage;
```
This was giving en-US on firefox and chrome. On Brave it was giving en-GB. For some reason when romcal is set to en-GB some of the romcal array does not show the data. I am guessing it might be because I may not have compiled the romcal with en-GB translation data.

```
        romcal.calendarFor({
            year: today.getFullYear() + i,
            scope: 'gregorian',
            country: 'singapore',
            locale: currentLocale,
            epiphanyOnSunday: true,
            corpusChristiOnSunday: true,
            ascensionOnSunday: false,
            outputOptionalMemorials: false,
            strictMode: true
```
Then changed it to fixed locale en-US and it seems to work now.
```
        romcal.calendarFor({
            year: today.getFullYear() + i,
            scope: 'gregorian',
            country: 'singapore',
            locale: 'en-US',
            epiphanyOnSunday: true,
            corpusChristiOnSunday: true,
            ascensionOnSunday: false,
            outputOptionalMemorials: false,
            strictMode: true
```
Not sure if this is the right thing to do but it seems to work.

Once again thank you very much for helping. I do appreciate it very much.

My bad for the late reply.

**Reply by tukusejssirs** [reply]
> The reason for the "-1" in the L192 is so that it will show the previous month as well. It is not perfect. Because in January it will not show December but that is ok.

In Javascript, the month in a date object can be only an integer from `0..11`. It cannot be a negative number. In order to make it work accross years, you either need to set some conditions yourself, or use some kind it package/library (e.g. [`dayjs`](https://day.js.org/), which supports [`substract()`](https://day.js.org/docs/en/manipulate/subtract) function).

> var currentLocale = navigator.language || navigator.userLanguage;

You can hard-code here the locale to any locale supported by romcal, e.g. `var currentLocale = 'en-US'` (choose you language) or directly in `locale: 'en-US'`, as you did. There is no need to use the browser locale. However, I believe you want to use Tamil, but romcal is currently not localised into Tamil, therefore even then it would fall back to `en` locale. If you want to contribute the translations (ideally based on the official books in Tamil, e.g. Roman Missal, lectionaries, Ordos, …), I’ll help you out with the technical stuff.

If you talk about `en-GB` not outputting all localised string, it might be a bug, as all string localised in `en-gb.ts`, they should be used, other strings should be used from `en.ts`.

**Reply by pvanthony** [reply]
Thank you again for the advice and help. I will checkout dayjs.

I will talk with Father Julian and see if we can get started with the Tamil and Sri Lankan translation.

Once again thank you for your help. It is has saved me much time and headache. :-)

**Reply by tukusejssirs** [reply]
You’re welcome. :wink:

I hope you two find some time to localise the strings in [`en.ts`](https://github.com/romcal/romcal/blob/dev/lib/locales/en.ts) to Tamil and Sri Lankan, at least those strings used by the calendar(s) used at your location, but ideally all the strings (that is the ideal for me: IMHO romcal should contain all strings localised into all supported languages, however, it is not necessary to localise them all in order to used them in calendar(s) usually localised in a particular locale).

**Reply by pvanthony** [reply]
Noted. Will ask for help if we get into trouble.

**Comment by pvanthony** [comment, marked as answer]
Marking this as answered.
